modulemd/v2/modulemd-module-stream.c

```diff
@@ -0,0 +1,346 @@
+
+enum
+{
+  PROP_0,
+  PROP_MDVERSION,
+  PROP_MODULE_NAME,
+  PROP_STREAM_NAME,
+  N_PROPS
+};
+
+static GParamSpec *properties[N_PROPS];
+
+ModulemdModuleStream *
+modulemd_module_stream_new (guint64 mdversion,
+                            const gchar *module_name,
+                            const gchar *stream_name)
+{
+  switch (mdversion)
+    {
+    case MD_MODULESTREAM_VERSION_ONE:
+      return MODULEMD_MODULE_STREAM (
+        modulemd_module_stream_v1_new (module_name, stream_name));
+
+    case MD_MODULESTREAM_VERSION_TWO:
+      return MODULEMD_MODULE_STREAM (
+        modulemd_module_stream_v2_new (module_name, stream_name));
+
+    default:
+      /* Other versions have not yet been implemented */
+      return NULL;
+    }
+}
+
+
+static void
+modulemd_module_stream_finalize (GObject *object)
+{
+  ModulemdModuleStream *self = (ModulemdModuleStream *)object;
+  ModulemdModuleStreamPrivate *priv =
+    modulemd_module_stream_get_instance_private (self);
+
+  g_clear_pointer (&priv->module_name, g_free);
+  g_clear_pointer (&priv->stream_name, g_free);
+
+  G_OBJECT_CLASS (modulemd_module_stream_parent_class)->finalize (object);
+}
+
+
+static ModulemdModuleStream *
+modulemd_module_stream_default_copy (ModulemdModuleStream *self,
+                                     const gchar *stream_name)
+{
+  const gchar *stream = NULL;
+
+  /* If the stream name was passed in, replace the original with it */
+  if (stream_name)
+    {
+      stream = stream_name;
+    }
+  else
+    {
+      stream = modulemd_module_stream_get_stream_name (self);
+    }
+
+  /* The base implementation of the copy */
+  return modulemd_module_stream_new (
+    modulemd_module_stream_get_mdversion (self),
+    modulemd_module_stream_get_module_name (self),
+    stream);
+}
+
+
+ModulemdModuleStream *
+modulemd_module_stream_copy (ModulemdModuleStream *self,
+                             const gchar *stream_name)
+{
+  ModulemdModuleStreamClass *klass;
+
+  g_return_val_if_fail (MODULEMD_IS_MODULE_STREAM (self), NULL);
+
+  if (!self)
+    return NULL;
+
+  g_return_val_if_fail (MODULEMD_IS_MODULE_STREAM (self), NULL);
+
+  klass = MODULEMD_MODULE_STREAM_GET_CLASS (self);
+  g_return_val_if_fail (klass->copy, NULL);
+
+  return klass->copy (self, stream_name);
+}
+
+
+ModulemdModuleStream *
+modulemd_module_stream_upgrade (ModulemdModuleStream *self,
+                                guint64 mdversion,
+                                GError **error)
+{
+  g_return_val_if_fail (MODULEMD_IS_MODULE_STREAM (self), NULL);
+  g_return_val_if_fail (
+    mdversion && mdversion <= MD_MODULESTREAM_VERSION_LATEST, NULL);
+  /* TODO */
+  return NULL;
+}
+
+
+static gboolean
+modulemd_module_stream_default_validate (ModulemdModuleStream *self,
+                                         GError **error)
+{
+  ModulemdModuleStreamPrivate *priv =
+    modulemd_module_stream_get_instance_private (self);
+
+  if (!priv->mdversion)
```

Comment:
Maybe we should just use the getter to call it on the child class?

Reply:
I reworked this so this check is properly done in the parent class.